clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1141,21 +1141,27 @@ private Map<TopicPartition, Long> beginningOrEndOffset(Collection<TopicPartition
             if (partitions.isEmpty()) {
                 return Collections.emptyMap();
             }
+
             Map<TopicPartition, Long> timestampToSearch = partitions
                 .stream()
                 .collect(Collectors.toMap(Function.identity(), tp -> timestamp));
             Timer timer = time.timer(timeout);
             ListOffsetsEvent listOffsetsEvent = new ListOffsetsEvent(
                 timestampToSearch,
-                false,
                 timer);
-            Map<TopicPartition, OffsetAndTimestamp> offsetAndTimestampMap = applicationEventHandler.addAndGet(
+
+            // shortcut the request if the timeout is zero.
+            if (timeout.isZero()) {
```

Comment:
Sorry, only noticed this now, but the original consumer seems to send a list offset request even if the timeout is 0, and you are specifically introducing code to avoid that. Isn't that going against what Kirk is trying to achieve? cc @kirktrue

Reply:
Hey @lucasbru - It seems like one of the handlers would also update the subscription state upon completion. See the snippet below:
```
                public void onSuccess(ListOffsetResult value) {
                    synchronized (future) {
                        result.fetchedOffsets.putAll(value.fetchedOffsets);
                        remainingToSearch.keySet().retainAll(value.partitionsToRetry);

                        offsetFetcherUtils.updateSubscriptionState(value.fetchedOffsets, isolationLevel);
                    }
                }

```

I think addAndGet seems to be sufficient to handle such logic so I'll revert this code.

Reply:
scratch off the previous comment - addAndGet actually doesn't. We will need to explicitly return an empty result.  See the code change.